instance/models/openedx_instance.py

```diff
@@ -223,7 +223,15 @@ def set_active_vm_dns_records(self):
             if ip_addr:
                 domain = "vm{index}.{base_domain}".format(index=i, base_domain=self.internal_lms_domain)
                 gandi.api.set_dns_record(domain, type="A", value=ip_addr)
-        # TODO: implement cleaning up DNS addresses that are no longer needed.
+
+    def clean_up_appservers_dns_records(self):
+        """
+        Removes the DNS records for app servers
+        """
+        self.logger.info("Cleaning up DNS records for app servers...")
+        for i, _ in enumerate(self.get_active_appservers(), 1):
```

Comment:
@pkulkark, there is a flaw in the logic used here (same as in the `set_active__vm_dns_records` method. Let's say I create a new instance, spawn an appserver and activate it. The appserver VM will get a DNS record set for `vm1.<instance domain>` on activation. Now there is a security update for which the instance gets redeployed and hence a new appserver is spawned. Before deactivating the 1st appserver which will cause some downtime, one would activate the 2nd appserver (which will then get the `vm2.<instance domain>` record set) and then deactivate the old one after the former is active. This will trigger a call to the `set_active_vm_dns_records` method which will re-assign the `vm1.<instance>` domain name to the 2nd appserver but won't do anything about the `vm2.<instance domain>` record previously assigned to the 2nd appserver.

The current implementation of the cleanup method will only remove the `vm1.<instance domain>` record and leave the `vm2.<instance domain>` record intact. This is not what we want. In scenarios where we have multiple active appservers at any point in time, this problem gets worse. So the logic in this method and the `set_active_vm_dns_records` method (and possibly, its callers) has to be updated to cleanup the records properly.

Reply:
@lgp171188 I changed the logic for `set_active_vm_dns_records` so that it only sets a DNS record while activating and takes the particular appserver number i.e. when we activate appserver 1, it will set `vm1.<instance domain>`. And calling the clean up method when we deactivate an appserver. Does that approach make more sense?